src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Flow.java

```diff
@@ -739,18 +736,108 @@ public void visitSwitchExpression(JCSwitchExpression tree) {
+                }
+            }
+            return constants;
+        }
+
+        private boolean coversDeconstructionStartingFromComponent(DiagnosticPosition pos,
+                                                                  Type targetType,
+                                                                  List<JCDeconstructionPattern> patterns,
+                                                                  int component) {
+            List<? extends RecordComponent> components = patterns.head.record.getRecordComponents();
+
+            if (components.size() == component) {
+                return true;
+            }
+
+            Type parameterizedComponentType = types.memberType(targetType, components.get(component));
+            List<JCPattern> nestedComponentPatterns = patterns.map(d -> d.nested.get(component));
+            Set<Symbol> nestedCovered = coveredSymbols(pos, parameterizedComponentType,
+                                                       nestedComponentPatterns);
+            Map<Symbol, List<JCDeconstructionPattern>> componentType2Patterns = new HashMap<>();
+            Set<Symbol> covered = new HashSet<>();
+
+            for (JCDeconstructionPattern subTypeCandidate : patterns) {
+                JCPattern nestedPattern = subTypeCandidate.nested.get(component);
+                Symbol currentPatternType;
+                switch (nestedPattern.getTag()) {
+                    case BINDINGPATTERN, PARENTHESIZEDPATTERN -> {
+                        PatternPrimaryType primaryPatternType =
+                                TreeInfo.primaryPatternType(nestedPattern);
+                        currentPatternType = primaryPatternType.type().tsym;
+                    }
+                    case DECONSTRUCTIONPATTERN -> {
+                        currentPatternType = ((JCDeconstructionPattern) nestedPattern).record;
+                    }
+                    default -> {
+                        throw Assert.error("Unexpected tree kind: " + nestedPattern.getTag());
+                    }
+                }
+                for (Symbol currentType : nestedCovered) {
+                    if (types.isSubtype(types.erasure(currentType.type),
+                                        types.erasure(currentPatternType.type))) {
+                        componentType2Patterns.put(currentType,
+                                                   componentType2Patterns.getOrDefault(currentType,
+                                                                                       List.nil())
+                                              .prepend(subTypeCandidate));
+                    }
+                }
+            }
+
+            for (Entry<Symbol, List<JCDeconstructionPattern>> e : componentType2Patterns.entrySet()) {
+                if (coversDeconstructionStartingFromComponent(pos, targetType, e.getValue(), component + 1)) {
+                    covered.add(e.getKey());
```

Comment:
So... it seems to me that what this code is doing is that, for a component index _i_, we get all its recursively covered symbols - then we add them to the set of covered symbols of component _i_, but only if components _w_, where _w_ > _i_ are also covered? That is, if in `R(P1, P2, ... Pn)`, we see that `P1` is covered, but `P2` is not, we also consider `P1` not covered (which in turn makes `R` not covered).